pkg/kubefed2/unjoin.go

```diff
@@ -0,0 +1,409 @@
+	opts := &unjoinFederation{}
+
+	cmd := &cobra.Command{
+		Use:     "unjoin CLUSTER_NAME --host-cluster-context=HOST_CONTEXT",
+		Short:   "Unjoin a cluster from a federation",
+		Long:    unjoin_long,
+		Example: unjoin_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := opts.Complete(args)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+
+			err = opts.Run(cmdOut, config)
+			if err != nil {
+				glog.Fatalf("error: %v", err)
+			}
+		},
+	}
+
+	flags := cmd.Flags()
+	opts.CommonBind(flags)
+	opts.Bind(flags)
+
+	return cmd
+}
+
+// Complete ensures that options are valid and marshals them if necessary.
+func (j *unjoinFederation) Complete(args []string) error {
+	err := j.SetName(args)
+	if err != nil {
+		return err
+	}
+
+	if j.clusterContext == "" {
+		glog.V(2).Infof("Defaulting cluster context to unjoining cluster name %s", j.ClusterName)
+		j.clusterContext = j.ClusterName
+	}
+
+	glog.V(2).Infof("Args and flags: name %s, host: %s, host-system-namespace: %s, kubeconfig: %s, cluster-context: %s, dry-run: %v",
+		j.ClusterName, j.Host, j.FederationNamespace, j.Kubeconfig, j.clusterContext, j.DryRun)
+
+	return nil
+}
+
+// Run is the implementation of the `unjoin federation` command.
+func (j *unjoinFederation) Run(cmdOut io.Writer, config util.FedConfig) error {
+	hostConfig, err := config.HostConfig(j.Host, j.Kubeconfig)
+	if err != nil {
+		// TODO(font): Return new error with this same text so it can be output
+		// by caller.
+		glog.V(2).Infof("Failed to get host cluster config: %v", err)
+		return err
+	}
+
+	clusterConfig, err := config.ClusterConfig(j.clusterContext, j.Kubeconfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get unjoining cluster config: %v", err)
+		return err
+	}
+
+	err = UnJoinCluster(hostConfig, clusterConfig, j.FederationNamespace,
+		j.Host, j.clusterContext, j.ClusterName, j.removeFromRegistry, j.DryRun)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UnJoinCluster performs all the necessary steps to unjoin a cluster from the
+// federation provided the required set of parameters are passed in.
+func UnJoinCluster(hostConfig, clusterConfig *rest.Config, federationNamespace, host,
+	unjoiningClusterContext, unjoiningClusterName string, removeFromRegistry, dryRun bool) error {
+
+	hostClientset, err := util.HostClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get host cluster clientset: %v", err)
+		return err
+	}
+
+	clusterClientset, err := util.ClusterClientset(clusterConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get unjoining cluster clientset: %v", err)
+		return err
+	}
+
+	fedClientset, err := util.FedClientset(hostConfig)
+	if err != nil {
+		glog.V(2).Infof("Failed to get federation clientset: %v", err)
+		return err
+	}
+
+	glog.V(2).Infof("Performing preflight checks for unjoin cluster: %s.", unjoiningClusterName)
+	err = performPreflightChecksForUnjoin(clusterClientset, unjoiningClusterName, host, federationNamespace)
```

Comment:
This method checks for the service account in the member cluster, and will fail the operation if it doesn't exist.  Yet removal of the cluster and federatedcluster will take place in advance of removal of that same service account.  If removal of the service account were to fail, the resources required to retry the operation would not be available (federated cluster referencing auth secret and cluster for api endpoint).  I don't think this represents a very effective approach to resiliency.  Consider removing the resources in the reverse order to maximize the chances of successful cleanup.

Reply:
Here I just want to use same logic to check of the member cluster was already joined to federation, if the serviceAccount exist, then I can say the check passed and we can go ahead with the unjoin operation.